File: Softwarechallenge-Tools/src/Evolution.java
import jargs.gnu.CmdLineParser;
import jargs.gnu.CmdLineParser.IllegalOptionValueException;
import jargs.gnu.CmdLineParser.UnknownOptionException;
import xerus.util.SysoutListener;
import xerus.util.tools.FileTools;
import xerus.util.tools.StringTools;
import xerus.util.tools.Tools;

import java.io.File;
import java.io.IOException;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.TimeUnit;

public class Evolution {

	private static String basepath;
	private static String ailoc;
	private static String datafile;
	private static String bestfile;

	private static boolean debug;

	public static void main(String[] args) throws IOException, InterruptedException, UnknownOptionException, IllegalOptionValueException {
		CmdLineParser parser = new CmdLineParser();
		CmdLineParser.Option option = parser.addIntegerOption('l', "line");
		CmdLineParser.Option path = parser.addStringOption('p', "path");
		CmdLineParser.Option debugOption = parser.addBooleanOption('d', "debug");
		parser.parse(args);

		basepath = (String) parser.getOptionValue(path, System.getProperty("user.dir") + "/");
		debug = (Boolean) parser.getOptionValue(debugOption, false);
		Process server = startServer();
		ailoc = basepath + "Elbdampfer3.4.jar";
		datafile = basepath + "strategies.csv";
		bestfile = basepath + "strategies_best.csv";
		FileTools.handle(e -> System.out.println(e.getMessage()));

		try {
			new Evolution((Integer) parser.getOptionValue(option, 0));
			while (true)
				new Evolution();
		} finally {
			server.destroy();
		}
	}

	private static Process startServer() throws IOException {
		ProcessBuilder serverbuilder = new ProcessBuilder(basepath + "testserver/start.bat");
		serverbuilder.directory(new File(basepath + "testserver"));
		serverbuilder.redirectErrorStream(true);
		return serverbuilder.start();
	}

	Strategy strategy;
	int line;

	public Evolution(int lineparam) throws InterruptedException {
		String[] file = FileTools.readall(datafile, 0, true).toArray(new String[0]);
		line = lineparam;
		if (lineparam == 0) {
			line = file.length;
			strategy = new Strategy(file[0]);
		} else {
			System.out.println("Lese von Zeile " + line);
			strategy = new Strategy(false, file[line]);
		}
		try {
			while (strategy.games < 200) {
				Process AI = startAIs();
				started = false;
				SysoutListener.addObserver(e -> {
					if (e.contains("Ich bin")) working();
				});
				Thread.sleep(6000);
				if (!started)
					buildAI().start();
				if (AI.waitFor(2, TimeUnit.MINUTES))
					strategy.write(AI.exitValue());
			}
			strategy.writeEnd("Finished");
		} catch (IOException e) {
			strategy.writeEnd("Error:" + e.getMessage());
		}
	}

	private boolean started;

	public void working() {
		started = true;
	}

	public Evolution() throws InterruptedException {
		this(0);
	}

	private Process startAIs() throws IOException {
		buildAI().start();
		ProcessBuilder pb = buildAI();
		if (debug)
			Collections.addAll(pb.command(), "-s", strategy.joinparams());
		else
			Collections.addAll(pb.command(), "-s", strategy.joinparams(), "-d", "0");
		return pb.inheritIO().start();
	}

	private static ProcessBuilder buildAI() {
		return new ProcessBuilder("java", "-jar", ailoc);
	}

	private class Strategy {
		double[] params;
		double[] variation;
		double winrate;
		int games;
		int won;
		int score;

		Strategy(String input) {
			this(true, input);
		}

		Strategy(boolean mutate, String info) {
			String[] infos = info.split(";");
			params = StringTools.split(infos[0]);
			variation = StringTools.split(infos[1]);
			if (mutate)
				mutate();
			else {
				c = 2;
				score = parseInt(infos);
				games = parseInt(infos);
				won = parseInt(infos);
			}
			write(0);
		}

		void mutate() {
			for (int i = 0; i < params.length; i++) {
				variation[i] = Tools.round((Math.random() * 2 - 1) * variation[i]);
				params[i] += variation[i];
			}
			games = 0;
			won = 0;
			score = 0;
		}

		int bestline;

		void write(int exitvalue) {
			if (exitvalue > 0) {
				games++;
				if (exitvalue / 100 >= 1)
					won++;
				score += exitvalue % 100;
			}

			String towrite = toString();
			if (FileTools.write(datafile, line, towrite))
				System.out.println(String.format("%s auf Zeile %s geschrieben ", towrite, line));
			if (games > 30) {
				if (winrate < 0.45)
					resetStrategy();
				if (games > 100) {
					if (winrate < 0.5) {
						resetStrategy();
					} else if (winrate > 0.52) {
						if (bestline == 0) {
							List<String> file = FileTools.readall(bestfile, 0, false);
							bestline = file.size();
						}
						FileTools.write(bestfile, bestline, towrite);
					}
				}
			}
		}

		void resetStrategy() {
			String file = null;
			while (file == null)
				file = FileTools.read(datafile, 0);
			strategy = new Strategy(file);
		}

		void writeEnd(String msg) {
			System.out.println("Fertig!");
			while (!FileTools.write(datafile, line, toString() + ";" + msg))
				Tools.sleep(10);
		}

		@Override
		public String toString() {
			winrate = Tools.round((double) won / games);
			return StringTools.join(joinparams(), StringTools.join(",", variation), "" + winrate, "" + score, "" + games, "" + won, Tools.time());
		}

		String joinparams() {
			return StringTools.join(",", params);
		}

		int c;

		private int parseInt(String[] s) {
			c++;
			if (s.length > c)
				return Integer.parseInt(s[c]);
			return 0;
		}

	}

}

File: softwarechallenge-tools/src/Evolution.kt
import jargs.gnu.CmdLineParser
import xerus.ktutil.*
import java.io.File
import java.util.concurrent.TimeUnit

object Evolution : EvolutionBase() {
	
	private const val GAMES = 100
	
	override lateinit var baseDir: File
	override lateinit var evolutionDir: File
	private lateinit var aiFile: File
	private lateinit var bestFile: File
	
	private var debug = false
	private var mutator = false
	
	private lateinit var server: Process
	private var running = true
		get() = server.isAlive && field
	
	@JvmStatic
	fun main(args: Array<String>) {
		val parser = CmdLineParser()
		val basedirOption = parser.addStringOption("basedir")
		val evolutiondirOption = parser.addStringOption("evolutiondir")
		val aiOption = parser.addStringOption("ai")
		val serverlocationOption = parser.addStringOption("server")
		val mutatorOption = parser.addBooleanOption("mutator")
		
		val debugOption = parser.addBooleanOption('d', "debug")
		val schoolOption = parser.addBooleanOption("school")
		val idOption = parser.addIntegerOption('i', "id")
		val portOption = parser.addStringOption('p', "port")
		
		parser.parse(args)
		
		mutator = parser.getValue(mutatorOption, false)
		school = parser.getValue(schoolOption, false)
		baseDir = parser.getValue(basedirOption, File(System.getProperty("user.dir"))) { File(it as String) }
		evolutionDir = File(baseDir, parser.getValue(evolutiondirOption, "evolution"))
		archiveDir.mkdirs()
		bestFile = evolutionDir.resolve("best.csv")
		
		port = parser.getValue(portOption, port)
		println("Port: $port")
		serverlocation = parser.getValue(serverlocationOption, if (school) "testserver/start.bat" else serverlocation)
		server = startServer()
		Runtime.getRuntime().addShutdownHook(Thread({
			running = false
			server.destroy()
			println("Stopping server...")
			server.waitFor()
		}, "Server terminator"))
		
		debug = parser.getValue(debugOption, false)
		aiFile = parser.getValue(aiOption, baseDir.resolve(if (school) "clients/start-client.bat" else "docker-client.sh")) { File(it as String) }.absoluteFile
		println("AI: $aiFile")
		
		var result = Evolve(parser.getValue(idOption) ?: getNextId()).start()
		while (result)
			result = Evolve().start()
	}
	
	private fun buildAI(): ProcessBuilder {
		val builder = if (aiFile.endsWith("jar"))
			ProcessBuilder("java", "-jar", aiFile.toString())
		else
			ProcessBuilder(aiFile.toString())
		builder.redirectOutput(logDir.resolve("${aiFile.name}-${System.currentTimeMillis()}.log"))
		if (debug)
			builder.redirectError(ProcessBuilder.Redirect.INHERIT)
		builder.directory(aiFile.parentFile)
		builder.command().addAll("--port", port)
		if (!debug)
			builder.command().addAll("-d", "0")
		return builder
	}
	
	class Evolve constructor(private val id: Int = getNextId()) {
		
		private val outputFile = file(id)
		private var strategy = if (outputFile.exists()) {
			println("Reading id $id")
			Strategy(file(id).readText(), false)
		} else {
			println("Got id $id")
			Strategy()
		}
		
		fun start(): Boolean {
			try {
				while (strategy.games < GAMES) {
					if (school) logDir.mkdirs()
					if (!running)
						return false
					val ai = startAI()
					if (ai.waitFor(1, TimeUnit.SECONDS) && ai.exitValue() != 0) {
						println("$ai exited unexpectedly with exit code ${ai.exitValue()}!")
						return false
					}
					if (ai.waitFor(3, TimeUnit.MINUTES)) {
						if (!running)
							return false
						val result = baseDir.resolve("clients/result$id")
						strategy.write(result.readText())
						result.delete()
					} else {
						println("The game did not finish in time!")
						return false
					}
					if (school) logDir.deleteRecursively()
				}
				strategy.writeEnd("Finished")
			} catch (e: Exception) {
				strategy.writeEnd("Error - $e")
				running = false
			}
			return running
		}
		
		private fun startAI(): Process {
			buildAI().start()
			val builder = buildAI()
			builder.command().addAll("-s", strategy.joinparams(), "-e", id.toString())
			return builder.redirectOutput(ProcessBuilder.Redirect.INHERIT).start()
		}
		
		internal fun resetStrategy() {
			outputFile.renameTo(archiveDir.resolve("$id-${System.currentTimeMillis()}"))
			println("Resetting!")
			strategy = Strategy()
		}
		
		private inner class Strategy internal constructor(input: String = bestFile.safe { readLines().let { if (mutator) it[(Math.random() * it.size).toInt()] else it[0] } }, mutate: Boolean = true) {
			internal var params: DoubleArray
			internal var variation: DoubleArray
			internal var winrate: Double = 0.0
			internal var games: Int = 0
			internal var won: Int = 0
			internal var score: Int = 0
			
			internal val bestLine = lazy(LazyThreadSafetyMode.NONE) { bestFile.readLines().size }
			
			init {
				val split = input.split(separator)
				params = split[0].split(',').map { it.toDouble() }.toDoubleArray()
				variation = split[1].split(',').map { it.toDouble() }.toDoubleArray()
				if (mutate)
					mutate()
				else {
					var c = 2
					fun parseInput(s: List<String>) = if (s.size > ++c) s[c].toInt() else 0
					score = parseInput(split)
					games = parseInput(split)
					won = parseInput(split)
				}
				write("-1 0")
			}
			
			internal fun mutate() {
				for (i in params.indices) {
					variation[i] = ((Math.random() * 2.2 - 1.1) * variation[i]).round()
					params[i] += variation[i]
				}
				games = 0
				won = 0
				score = 0
			}
			
			internal fun write(result: String) {
				val split = result.split(' ')
				if (split[0].toInt() >= 0) {
					games++
					if (split[0] == "1")
						won++
				}
				score += split[1].toInt()
				
				val s = toString()
				outputFile.writeText(s)
				println("Wrote \"$s\" to $outputFile")
				if (games > 15) {
					val bcd = binominalCD(won, games)
					if (bcd < 0.4)
						resetStrategy()
					if (bestLine.isInitialized() || (games > 50 && bcd > 0.95)) {
						bestFile.safe {
							println("Good Strategy - Writing to $bestFile line ${bestLine.value}")
							write(bestLine.value, s)
						}
					}
				}
			}
			
			internal fun writeEnd(msg: String) {
				println("Done: $msg")
				println()
				outputFile.writeText(toString() + separator + msg)
			}
			
			override fun toString(): String {
				winrate = won.toDouble() / games
				return arrayOf(joinparams(), variation.joinToString(",") { it.format(2) }, winrate.format(2), score, games, won, formattedTime()).joinToString(separator)
			}
			
			internal fun joinparams() = params.joinToString(",") { it.format(2) }
			
		}
		
	}
	
}